src/lib/device-online-state/index.ts

```diff
@@ -0,0 +1,282 @@
+			}
+
+			return pollInterval.value;
+		})
+		.catch(PollIntervalNotFoundError, () => {
+			// query to determine the poll interval for this device...
+			return resinApi
+				.get({
+					resource: 'application_config_variable',
+					passthrough: {
+						req: root,
+					},
+					options: {
+						$select: 'value',
+						$top: 1,
+						$filter: {
+							application: {
+								$any: {
+									$alias: 'a',
+									$expr: {
+										a: {
+											owns__device: {
+												$any: {
+													$alias: 'd',
+													$expr: {
+														d: {
+															uuid: uuid,
+														},
+													},
+												},
+											},
+										},
+									},
+								},
+							},
+							name: 'RESIN_SUPERVISOR_POLL_INTERVAL',
+						},
+					},
+				})
+				.then(([pollInterval]: Array<{ value: number }>) => {
+					if (pollInterval) {
+						return pollInterval.value;
+					}
+
+					return DEFAULT_SUPERVISOR_POLL_INTERVAL;
+				});
+		})
+		.then(pollInterval => {
+			return Math.max(pollInterval, DEFAULT_SUPERVISOR_POLL_INTERVAL);
+		});
+};
+
+export enum DeviceOnlineStates {
+	Unknown = -2,
+	Timeout = -1,
+	Offline = 0,
+	Online = 1,
+}
+
+class NoMessagesToConsumeSignal extends Error {}
+
+class DeviceOnlineStateManager extends EventEmitter {
+	private static readonly REDIS_NAMESPACE = 'device-online-state';
+	private static readonly EXPIRED_QUEUE = 'expired';
+
+	rsmq: RedisSMQ;
+	redis: PromisifedRedisClient;
+
+	constructor() {
+		super();
+
+		// create a new Redis client...
+		this.redis = createPromisifedRedisClient({
+			host: REDIS_HOST,
+			port: REDIS_PORT,
+		});
+
+		// initialise the RedisSMQ object using our Redis client...
+		this.rsmq = new RedisSMQ({
+			client: this.redis,
+			ns: DeviceOnlineStateManager.REDIS_NAMESPACE,
+		});
+
+		// create the RedisMQ queue and start consuming messages...
+		this.rsmq
+			.createQueueAsync({ qname: DeviceOnlineStateManager.EXPIRED_QUEUE })
+			.catch((err: Error) => {
+				if (err.name !== 'queueExists') {
+					throw err;
+				}
+			})
+			.then(() => {
+				this.consume();
+			});
+	}
+
+	// event emitters
+	private deviceOnline(uuid: string) {
+		this.emit('change', uuid, DeviceOnlineStates.Online);
+	}
+
+	private deviceOffline(uuid: string) {
+		this.emit('change', uuid, DeviceOnlineStates.Offline);
+	}
+
+	private deviceTimeout(uuid: string) {
+		this.emit('change', uuid, DeviceOnlineStates.Timeout);
+	}
+
+	private consume() {
+		// pull a message from the queue...
+		this.rsmq
+			.receiveMessageAsync({
+				qname: DeviceOnlineStateManager.EXPIRED_QUEUE,
+				vt: 30, // prevent other consumers seeing the same message (if any) for 30 secs (prevents multiple API agents from processing it)...
+			})
+			.tap((msg: any) => {
+				// no message! this is thrown to restart the consume process...
+				if (!msg || !msg.id) {
+					return Bluebird.delay(1000).thenThrow(
+						new NoMessagesToConsumeSignal(),
+					);
+				}
+			})
+			.then((msg: any) => {
```

Comment:
This whole block can be simplified down to something like this:

```js
.then(({ id, message }) => {
	const { uuid, nextState } = JSON.parse(message);
	switch (nextState) {
		// ...
	}
	return this.rsmq.deleteMessageAsync({
		// ...
	});
})
```

Any errors thrown by `JSON.parse` or from event handlers will be caught by the next catch handler.